### PR TITLE
Added support for hiding notes from install/upgrade output

### LINKED ISSUE #11598: FR: `--skip-chart-notes` for install/upgrade
Refer to %subject% :-)  
(Or just `--skip-notes`?)
 
TL;DR

```
% helm install ...
NAME: ingress-nginx
LAST DEPLOYED: Sat Dec  3 18:16:12 2022
NAMESPACE: ingress-nginx
STATUS: deployed
REVISION: 4
TEST SUITE: None
NOTES:
The ingress-nginx controller has been installed.
It may take a few minutes for the LoadBalancer IP to be available.
You can watch the status by running 'kubectl --namespace ingress-nginx get services -o wide -w ingress-nginx-controller'
...
%
```

VERSUS (feature request)

```
% helm install --skip-chart-notes ...
NAME: ingress-nginx
LAST DEPLOYED: Sat Dec  3 18:16:12 2022
NAMESPACE: ingress-nginx
STATUS: deployed
REVISION: 4
TEST SUITE: None
%
```

### PR DESCRIPTION
**Pull request by miles-w-3**
Signed-off-by: Miles Wilson <wilson.mil@icloud.com>

**What this PR does / why we need it**:
Closes #11598, allows for notes to be hidden during chart installation, but doesn't affect their preservation in chart metadata, since that seems like a larger change that might have unintended consequences.

**Special notes for your reviewer**:
This is my first helm commit, if there are practices I missed or larger considerations I should be making, please let me know. I'm looking forward to working more on the source code for helm.

**If applicable**:
- [ ] this PR contains documentation
- [ ] this PR contains unit tests
- [X] this PR has been tested for backwards compatibility

### CODE REVIEW

**Review by joejulian — APPROVED** [review]
Works for me

**Comment by joejulian** [comment]
@miles-w-3 Can you rebase this?

**Comment by miles-w-3** [comment]
Sure I can work on it now, looks like there are additions to statusPrinter in the same place which is causing the conflict:
```
type statusPrinter struct {
	release         *release.Release
	debug           bool
	showDescription bool
<<<<<<< HEAD
	showResources   bool
=======
	skipNotes       bool
>>>>>>> 5cc99712 (Added support for hiding notes from install/upgrade output)
}

```

File: cmd/helm/install.go
Comment (joejulian) [inline cmd/helm/install.go:161]: Oops, you included the merge conflicts.

Comment (miles-w-3) [reply cmd/helm/install.go]: The problem is that there are additional conflicts that aren't being picked up, where both branches have `&statusPrinter{rel, settings.Debug, false, false}`, so they all need to be amended. I'll keep working on this. I've preserved the original branch as `old-chart-notes` in my fork. The current branch is free of merge conflicts but will not build.

Comment (miles-w-3) [reply cmd/helm/install.go]: @joejulian finally got it, all of the changes should be reconciled

Comment (joejulian) [reply cmd/helm/install.go]: Thank you

File: cmd/helm/status.go
Comment (joejulian) [inline cmd/helm/status.go:83]: should this be

```suggestion
			return outfmt.Write(out, &statusPrinter{rel, false, client.ShowDescription, client.ShowResources, false})
```
or maybe even add HideNotes to the status command so you can use client.HideNotes in there.

File: cmd/helm/upgrade.go
Comment (joejulian) [inline cmd/helm/upgrade.go:254]:
```suggestion
	f.BoolVar(&client.HideNotes, "hide-notes", false, "if set, do not show notes in install output. Does not affect presence in chart metadata")
```

File: cmd/helm/status.go
Comment (joejulian) [inline cmd/helm/status.go:115]: Should this be hideNodes to be consistent with the flag?

**Review by joejulian — CHANGES_REQUESTED** [review]
See in-line comments

**Comment by miles-w-3** [comment]
@joejulian thanks for your patience and for those corrections, I believe I fixed them all and rebased off of master again to resolve the new change conflicts.

**Review by joejulian — DISMISSED** [review]
I just tested this. Install worked as expected but upgrade printed, "false" and the notes.

```
bin/helm upgrade --install -n rn redpanda redpanda/redpanda --create-namespace --set=statefulset.replicas=1 --hide-notes --set external.domain=my.foo.bar
false
Release "redpanda" has been upgraded. Happy Helming!
NAME: redpanda
LAST DEPLOYED: Tue Aug 29 16:35:29 2023
NAMESPACE: rn
STATUS: deployed
REVISION: 2
NOTES:
Congratulations on installing redpanda!

The pods will rollout in a few seconds. To check the status:

  kubectl -n rn rollout status statefulset redpanda --watch

Try some sample commands:

Get the api status:

  kubectl -n rn exec -ti redpanda-0 -c redpanda -- rpk cluster info --brokers redpanda-0.redpanda.rn.svc.cluster.local.:9093 --tls-enabled --tls-truststore /etc/tls/certs/default/ca.crt 

Create a topic

  kubectl -n rn exec -ti redpanda-0 -c redpanda -- rpk topic create test-topic --brokers redpanda-0.redpanda.rn.svc.cluster.local.:9093 --tls-enabled --tls-truststore /etc/tls/certs/default/ca.crt 

Describe the topic:

  kubectl -n rn exec -ti redpanda-0 -c redpanda -- rpk topic describe test-topic --brokers redpanda-0.redpanda.rn.svc.cluster.local.:9093 --tls-enabled --tls-truststore /etc/tls/certs/default/ca.crt 

Delete the topic:

  kubectl -n rn exec -ti redpanda-0 -c redpanda -- rpk topic delete test-topic --brokers redpanda-0.redpanda.rn.svc.cluster.local.:9093 --tls-enabled --tls-truststore /etc/tls/certs/default/ca.crt
```

**Comment by miles-w-3** [comment]
Working on figuring out where the false is being printed from, then I will debug the behavior

**Comment by miles-w-3** [comment]
@joejulian it looks like the `false` print comes from this change that has been committed to main: https://github.com/helm/helm/blob/387ba9c5b7431aea4af0732c465ebbcbd18b5805/pkg/action/upgrade.go#L265

I have run through all of the test cases, I believe all cases are addressed except for that `false` print. Let me know if you want me to remove it.
```bash
# my built binary
❯ which delm
delm: aliased to /Users/mileswilson/repos/open-source/helm/bin/helm

# pure install, no notes
❯ delm install test helm-test --hide-notes
NAME: test
LAST DEPLOYED: Fri Sep  1 00:04:43 2023
NAMESPACE: default
STATUS: deployed
REVISION: 1

# upgrade install with existing chart - falls back to upgrade logic
❯ delm upgrade --install test helm-test --hide-notes
false
Release "test" has been upgraded. Happy Helming!
NAME: test
LAST DEPLOYED: Fri Sep  1 00:05:02 2023
NAMESPACE: default
STATUS: deployed
REVISION: 2

# upgrade on installed chart
❯ delm upgrade test helm-test --hide-notes
false
Release "test" has been upgraded. Happy Helming!
NAME: test
LAST DEPLOYED: Fri Sep  1 00:05:32 2023
NAMESPACE: default
STATUS: deployed
REVISION: 3

❯ helm uninstall test
release "test" uninstalled

# upgrade --install, uses clean install - goes into additional upgrade case
❯ delm upgrade --install test helm-test --hide-notes
Release "test" does not exist. Installing it now.
NAME: test
LAST DEPLOYED: Fri Sep  1 00:06:02 2023
NAMESPACE: default
STATUS: deployed
REVISION: 1
```

FWIW, when I run on 3.11 i don't see this:
```
❯ helm version
version.BuildInfo{Version:"v3.11.1", GitCommit:"293b50c65d4d56187cd4e2f390f0ada46b4c4737", GitTreeState:"clean", GoVersion:"go1.19.5"}

❯ helm upgrade test helm-test
Release "test" has been upgraded. Happy Helming!
NAME: test
LAST DEPLOYED: Fri Sep  1 00:12:39 2023
NAMESPACE: default
STATUS: deployed
REVISION: 2
NOTES:
1. Get the application URL by running these commands:
  export POD_NAME=$(kubectl get pods --namespace default -l "app.kubernetes.io/name=helm-test,app.kubernetes.io/instance=test" -o jsonpath="{.items[0].metadata.name}")
  export CONTAINER_PORT=$(kubectl get pod --namespace default $POD_NAME -o jsonpath="{.spec.containers[0].ports[0].containerPort}")
  echo "Visit http://127.0.0.1:8080 to use your application"
  kubectl --namespace default port-forward $POD_NAME 8080:$CONTAINER_PORT
```

But when I run built from main:
```
❯ delm version
version.BuildInfo{Version:"v3.12+unreleased", GitCommit:"387ba9c5b7431aea4af0732c465ebbcbd18b5805", GitTreeState:"dirty", GoVersion:"go1.20.1"}

❯ delm upgrade --install test helm-test
false
Release "test" has been upgraded. Happy Helming!
NAME: test
LAST DEPLOYED: Fri Sep  1 00:15:34 2023
NAMESPACE: default
STATUS: deployed
REVISION: 4
NOTES:
1. Get the application URL by running these commands:
  export POD_NAME=$(kubectl get pods --namespace default -l "app.kubernetes.io/name=helm-test,app.kubernetes.io/instance=test" -o jsonpath="{.items[0].metadata.name}")
  export CONTAINER_PORT=$(kubectl get pod --namespace default $POD_NAME -o jsonpath="{.spec.containers[0].ports[0].containerPort}")
  echo "Visit http://127.0.0.1:8080 to use your application"
  kubectl --namespace default port-forward $POD_NAME 8080:$CONTAINER_PORT
```
I also get the print statement.
Sorry for all of the runaround on this.

**Comment by shoce** [comment]
Thanks for making this! Waiting for this PR to be merged!

**Comment by joejulian** [comment]
@miles-w-3 needs rebased again and I have a request. If it's not too much trouble, could this flag also work for `helm test`? I have been realizing the last few days (as I've added detailed instructions for things to notes) that my test results are getting scrolled off the screen after the tests complete.

**Comment by miles-w-3** [comment]
@joejulian Rebased and added functionality to `helm test`, here is an example where the chart is also called test

```
❯ delm test test
NAME: test
LAST DEPLOYED: Thu Oct  5 15:57:00 2023
NAMESPACE: default
STATUS: deployed
REVISION: 1
TEST SUITE:     test-helm-test-test-connection
Last Started:   Thu Oct  5 15:57:05 2023
Last Completed: Thu Oct  5 15:57:09 2023
Phase:          Succeeded
NOTES:
1. Get the application URL by running these commands:
  export POD_NAME=$(kubectl get pods --namespace default -l "app.kubernetes.io/name=helm-test,app.kubernetes.io/instance=test" -o jsonpath="{.items[0].metadata.name}")
  export CONTAINER_PORT=$(kubectl get pod --namespace default $POD_NAME -o jsonpath="{.spec.containers[0].ports[0].containerPort}")
  echo "Visit http://127.0.0.1:8080 to use your application"
  kubectl --namespace default port-forward $POD_NAME 8080:$CONTAINER_PORT
```

```
❯ delm test test --hide-notes
NAME: test
LAST DEPLOYED: Thu Oct  5 15:57:00 2023
NAMESPACE: default
STATUS: deployed
REVISION: 1
TEST SUITE:     test-helm-test-test-connection
Last Started:   Thu Oct  5 15:58:08 2023
Last Completed: Thu Oct  5 15:58:11 2023
Phase:          Succeeded
```

FYI @shoce

**Comment by shoce** [comment]
it looks like ready for merge

**Review by joejulian — APPROVED** [review]

**Comment by butuzov** [comment]
C'mon Guys! Just one more approval? This goes on for 10 month!

**Comment by shoce** [comment]
is the one received approval enough for the merge?

**Comment by jbilliau-rcd** [comment]
Is this still an active PR? I would love to get this merged, looking for this option myself to make our helm tests output way less verbose with information we don't care about.

**Comment by shoce** [comment]
can someone finally merge this please?

**Comment by miles-w-3** [comment]
@joejulian I see this has been approved, are there any additional steps required before it can be merged?

**Review by sabre1041 — COMMENTED** [review]
Tested this new capability and it works as expected

However, I am wondering what the expectation would be if a user specified both the `--hide-notes` and `--render-subchart-notes` options. Currently, it does not print any notes as it is being suppressed by the `--hide-notes` option.  However, is there an expectation that one would want subchart notes to render without parent?

**Comment by miles-w-3** [comment]
I am happy to change the logic to support this, however if the driver of the request is to remove the verbosity from the chart installation output, I believe it would be beneficial to hide all notes

**Comment by shoce** [comment]
> Tested this new capability and it works as expected
> 
> However, I am wondering what the expectation would be if a user specified both the `--hide-notes` and `--render-subchart-notes` options. Currently, it does not print any notes as it is being suppressed by the `--hide-notes` option. However, is there an expectation that one would want subchart notes to render without parent?

No, there is no expectation to see subchart notes with --hide-notes option. Though giving an error or a warning would be helpful if both --hide-notes and --render-subchart-notes options are provided.

**Comment by miles-w-3** [comment]
@shoce I can make that change as well. But one thing that's worth mentioning, `hide-notes` hides the chart note output, but the rendered output is still associated with the installation and can be queried from an installed instance on the cluster. For this reason, I don't think there is a conflict between these fields. If `--render-subchart-notes` is set, then subchart notes are still queryable from the install. But no matter what, when `--hide-notes` is set, nothing notes-related gets printed to the console

**Comment by shoce** [comment]
> @shoce I can make that change as well. But one thing that's worth mentioning, `hide-notes` hides the chart note output, but the rendered output is still associated with the installation and can be queried from an installed instance on the cluster. For this reason, I don't think there is a conflict between these fields. If `--render-subchart-notes` is set, then subchart notes are still queryable from the install. But no matter what, when `--hide-notes` is set, nothing notes-related gets printed to the console

I absolutely agree but still I do not have any permissions on this repo to merge so will need to wait for the contributors.

**Comment by shoce** [comment]
this is one of that cases when while waiting for the merge i realized i can live without it. not looking like moving anywhere.

**Comment by sabre1041** [comment]
> I am happy to change the logic to support this, however if the driver of the request is to remove the verbosity from the chart installation output, I believe it would be beneficial to hide all notes

Thanks for the explanation and it makes sense. To enable moving forward, lets integrate this enhancement

**Review by sabre1041 — APPROVED** [review]
LGTM

**Pull request merged**
